packages/web3/src/codec/asset-output-codec.ts

```diff
@@ -0,0 +1,150 @@
+    })
+    .buffer('lockTime', {
+      length: 8
+    })
+    .nest('tokens', {
+      type: tokensCodec.parser
+    })
+    .nest('additionalData', {
+      type: byteStringCodec.parser
+    })
+
+  encode(input: AssetOutput): Buffer {
+    const amount = Buffer.from(compactUnsignedIntCodec.encode(input.amount))
+    const lockupScript = lockupScriptCodec.encode(input.lockupScript)
+    const lockTime = Buffer.from(input.lockTime)
+    const tokens = Buffer.from(tokensCodec.encode(input.tokens.value))
+    const additionalData = Buffer.from(byteStringCodec.encode(input.additionalData))
+
+    return Buffer.concat([amount, lockupScript, lockTime, tokens, additionalData])
+  }
+
+  decode(input: Buffer): AssetOutput {
+    return this.parser.parse(input)
+  }
+
+  static toFixedAssetOutputs(txIdBytes: Uint8Array, outputs: AssetOutput[]): FixedAssetOutput[] {
+    return outputs.map((output, index) => AssetOutputCodec.toFixedAssetOutput(txIdBytes, output, index))
+  }
+
+  static toFixedAssetOutput(txIdBytes: Uint8Array, output: AssetOutput, index: number): FixedAssetOutput {
+    const attoAlphAmount = compactUnsignedIntCodec.toU256(output.amount).toString()
+    const lockTime = Number(longCodec.decode(output.lockTime))
+    const tokens = output.tokens.value.map((token) => {
+      return {
+        id: token.tokenId.toString('hex'),
+        amount: compactUnsignedIntCodec.toU256(token.amount).toString()
+      }
+    })
+    const message = output.additionalData.value.toString('hex')
+    const scriptType = output.lockupScript.scriptType
+    const key = binToHex(blakeHash(Buffer.concat([txIdBytes, signedIntCodec.encode(index)])))
+    const outputLockupScript = output.lockupScript.script
+    const address = bs58.encode(lockupScriptCodec.encode(output.lockupScript))
+
+    let hint: number | undefined = undefined
+    if (scriptType === 0) {
+      // P2PKH
+      hint = createHint((outputLockupScript as PublicKeyHash).publicKeyHash)
+    } else if (scriptType === 1) {
+      // P2MPKH
+      hint = createHint((outputLockupScript as MultiSig).publicKeyHashes.value[0].publicKeyHash)
+    } else if (scriptType === 2) {
+      // P2SH
+      hint = createHint((outputLockupScript as P2SH).scriptHash)
+    } else if (scriptType === 3) {
+      // P2C
+      hint = createHint((outputLockupScript as P2C).contractId)
+    } else {
+      throw new Error(`TODO: decode output script type: ${scriptType}`)
+    }
+
+    return { hint, key, attoAlphAmount, lockTime, tokens, address, message }
+  }
+
+  static fromFixedAssetOutputs(fixedOutputs: FixedAssetOutput[]): AssetOutput[] {
+    return fixedOutputs.map((output) => {
+      return AssetOutputCodec.fromFixedAssetOutput(output)
+    })
+  }
+
+  static fromFixedAssetOutput(fixedOutput: FixedAssetOutput): AssetOutput {
+    const amount: DecodedCompactInt = compactUnsignedIntCodec.decode(
+      compactUnsignedIntCodec.encodeU256(BigInt(fixedOutput.attoAlphAmount))
```

Comment:
After encoding, we decoded once again. Is it possible to directly convert bigint to `DecodedCompactInt`?

Reply:
`compactUnsignedIntCodec.decode` should be pretty efficient since it just takes the first byte, and store rest of the bytes into a `UintArray`. I can create a method to do `compactUnsignedIntCodec.decode(BigInt(fixedOutput.attoAlphAmount))`, WDYT?